### PR TITLE
fusion segmenter forward input to exclude uop used by SegmenterSet

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes: #2658

forwarded unary operation on SegmenterSet could potentially put computation ops in no-op segments.

This PR prevents any uop with output consumed by SegmenterSet to be forwarded.

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
Not sure who's a good reviewer for segmenter related issues.

**Comment by jjsjann123** [comment]
!build

File: tests/cpp/test_segmentation.cpp [hunk 1]
```diff
@@ -702,4 +704,29 @@ TEST_F(SegmentationTest, MultipleSegmentSetsInOneSegment) {
...
+  // Disable mark aliases prepare pass, which might insert more segment_set
+  preseg_passes::OptimizationPassGuard<preseg_passes::MarkAliasesPreparePass>
+      optimization_guard(false);
+
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* vanilla_in = makeContigConcreteTensor({2, 3});
+  TensorView* in = relu(vanilla_in);
+  TensorView* seg_in = segment_set(in);
```
Comment (jjsjann123) [inline tests/cpp/test_segmentation.cpp:717]: In issue #2658 , we ended up forwarding the `relu` op into `segment_set`. Ended up having two pointwise kernel.

**Review by jacobhinkle — APPROVED** [review]
LGTM. This seems safe to me. I am not sure it will 100% prevent this issues, since we could also naturally segment at a location where a NoOp segment then gets some forwarded uops. I think ideally we would actually include the forwarded ops in the canSchedule checks also.

**Comment by jjsjann123** [comment]
> we could also naturally segment at a location where a NoOp segment then gets some forwarded uops.

That sounds more like a segmenter issue in general. i.e. when should we try to get a no-op segment. Our current greedy segmenter doesn't consider that at all.

Hence @wujingyue 's mark alias pass tries to use `segmenter_set` to force a segment to facilitate that. At least with this PR, that attempt is not going to be undo by forwarded inputs. :crossed_fingers:

**Pull request merged**